### PR TITLE
Migrate ensure_executing transitions to new WorkerState event mechanism - part 1

### LINKED ISSUE #5895: Migrate `ensure_computing` transitions to new `WorkerState` event mechanism
The task executing transitions should be migrated to the `WorkerState` event mechanism as outlined in https://github.com/dask/distributed/issues/5736#issuecomment-1040599751

- The `Worker.execute` method is modified such that it no longer performs any transition but instead returns appropriate `StateMachineEvents` that trigger the necessary handlers. For instance
  - TaskFinished
  - Rescheduled
  - TaskErred
- The `ensure_computing` method is removed from the list of `handle_comm` `every_cycle` callbacks
- The logic of `ensure_computing` is moved to a private method (perspectively a method of the new class `WorkerState`)
- The new private method will not perform any transitions but rather return a list of recommendations
- This private method is then called as part of the transition system, e.g.
  - `unpause handler`
  - `transition_waiting_ready`
  - `transition_executing_*`
- The implicit goal is to remove all invocations of `self.loop.add_callback(self.execute, ...)` and replace this with the new callback method

### Blocked by 
- [x] https://github.com/dask/distributed/issues/5894

### Related work 
- [x] https://github.com/dask/distributed/issues/5891
- [ ] https://github.com/dask/distributed/issues/5896

### PR DESCRIPTION
**Pull request by crusaderky**
Partially closes #5895
See comment below for scope of this PR

### CODE REVIEW

File: distributed/worker.py [hunk 1]
```diff
@@ -2592,9 +2594,25 @@ def _handle_instructions(self, instructions: list[Instruction]) -> None:
         for inst in instructions:
             if isinstance(inst, SendMessageToScheduler):
                 self.batched_stream.send(inst.to_dict())
+            elif isinstance(inst, Execute):
+                self.loop.add_callback(
+                    self._async_instruction_callback,
+                    self.execute(inst.key, stimulus_id=inst.stimulus_id),
+                    stimulus_id=inst.stimulus_id,
+                )
```
Comment (crusaderky) [inline distributed/worker.py:2602]: This differs from the design document, which instead was creating a fire-and-forget task.
As far as I understand, the difference between the two is purely cosmetic.

Added value would be given from spawning a task, track it e.g. in a set ``Worker.running_asyncio_tasks``, and then cancel it in ``Worker.close()``. Even if desirable, however, I think this is best left to a future PR.

File: distributed/worker.py [hunk 2]
```diff
@@ -3499,30 +3516,16 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
...
-                **emsg,
-                stimulus_id=stimulus_id,
-            )
-        finally:
-            self.ensure_computing()
-            self.ensure_communicating()
+            msg = error_message(exc)
+            recommendations = {ts: tuple(msg.values())}
+
+        return recommendations, []
```
Comment (crusaderky) [inline distributed/worker.py:3528]: This differs from https://github.com/dask/distributed/issues/5895:

> - The `Worker.execute` method is modified such that it no longer performs any transition but instead returns appropriate `StateMachineEvents` that trigger the necessary handlers. For instance
>   - TaskFinished
>   - Rescheduled
>   - TaskErred

I could not find any benefit in implementing those events vs. just returning recommendations?

File: distributed/worker.py [hunk 3]
```diff
@@ -3412,17 +3429,22 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
 
             args2, kwargs2 = self._prepare_args_for_execution(ts, args, kwargs)
 
-            if ts.annotations is not None and "executor" in ts.annotations:
-                executor = ts.annotations["executor"]
-            else:
+            try:
+                executor = ts.annotations["executor"]  # type: ignore
+            except (TypeError, KeyError):
                 executor = "default"
```
Comment (fjetter) [inline distributed/worker.py:3435]: This is a change in behaviour that is not related to the state machine. I would prefer doing this in a separate PR.

File: distributed/worker.py (hunk 2 above)
Comment (fjetter) [reply distributed/worker.py:3533]: The benefit is that we want to log the events and keep the recommendations as an internal detail of the to-be-defined `WorkerState` class. Not using recommendations here is one of the more important points of the design proposals

File: distributed/worker.py [hunk 4]
```diff
@@ -1773,14 +1782,7 @@ def handle_cancel_compute(self, key: str, stimulus_id: str) -> None:
...
         """
-        ts = self.tasks.get(key)
-        if ts and ts.state in READY | {"waiting"}:
-            self.log.append((key, "cancel-compute", stimulus_id, time()))
-            # All possible dependents of TS should not be in state Processing on
-            # scheduler side and therefore should not be assigned to a worker,
-            # yet.
-            assert not ts.dependents
-            self.transition(ts, "released", stimulus_id=stimulus_id)
+        self.handle_stimulus(CancelComputeEvent(key=key, stimulus_id=stimulus_id))
```
Comment (crusaderky) [inline distributed/worker.py:1785]: In the future we should consider sending event objects directly from the scheduler

File: distributed/worker.py (hunk 2 above)
Comment (crusaderky) [reply distributed/worker.py]: Please have a look now

File: distributed/worker.py (hunk 4 above)
Comment (fjetter) [reply distributed/worker.py:1782]: Yes, I could see this very nicely being integrated in our RPC framework 👍

File: distributed/worker.py [hunk 5]
```diff
@@ -3392,24 +3414,16 @@ def ensure_computing(self) -> None:
...
-            return
-        if key not in self.tasks:
-            return
-        ts = self.tasks[key]
+            return None
+        ts = self.tasks.get(key)
+        if not ts:
+            return None
+        if ts.state == "cancelled":
+            return CancelComputeEvent(key=ts.key, stimulus_id=stimulus_id)
```
Comment (fjetter) [inline distributed/worker.py:3424]: I would suggest a name like `AlreadyCancelledEvent` to distinguish this from `Client.cancel` events passed through the scheduler. We don't have this, yet, but there is some potential for confusion

File: distributed/worker.py [hunk 6]
```diff
@@ -3468,75 +3480,126 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
...
+                self.active_keys.discard(key)
+
+            self.threads[key] = result["thread"]
+
             if result["op"] == "task-finished":
-                ts.nbytes = result["nbytes"]
-                ts.type = result["type"]
-                recommendations[ts] = ("memory", value)
                 if self.digests is not None:
                     self.digests["task-duration"].add(result["stop"] - result["start"])
```
Comment (fjetter) [inline distributed/worker.py:3489]: Same here. Do we want to move this to the result/event handler?

File: distributed/worker.py [hunk 7]
```diff
@@ -3468,75 +3480,126 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
...
-            emsg.pop("status")
-            self.transition(
-                ts,
+
+    @functools.singledispatchmethod
+    def handle_event(self, ev: StateMachineEvent) -> tuple[Recs, Instructions]:
+        raise TypeError(ev)  # pragma: nocover
+
+    # TODO Set return type annotation of all handle_event implementations
+    #      to tuple[Recs, Instructions] (requires Python >=3.9)
```
Comment (fjetter) [inline distributed/worker.py:3546]: Is this a functools limitation or why do we need py3.9? Do we even need individual annotations if they are all the same?

File: distributed/worker.py [hunk 8]
```diff
@@ -3468,75 +3480,126 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
...
+    @handle_event.register
+    def _(self, ev: CancelComputeEvent):  # -> tuple[Recs, Instructions]:
+        ts = self.tasks.get(ev.key)
```
Comment (fjetter) [inline distributed/worker.py:3549]: The tasks must always exist. There is no way this could've been dropped earlier. If so, that would be a severe bug

File: distributed/worker.py [hunk 9]
```diff
@@ -3468,75 +3480,126 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
...
-                ts,
+
+    @functools.singledispatchmethod
+    def handle_event(self, ev: StateMachineEvent) -> tuple[Recs, Instructions]:
+        raise TypeError(ev)  # pragma: nocover
+
+    # TODO Set return type annotation of all handle_event implementations
+    #      to tuple[Recs, Instructions] (requires Python >=3.9)
+    @handle_event.register
+    def _(self, ev: CancelComputeEvent):  # -> tuple[Recs, Instructions]:
```
Comment (fjetter) [inline distributed/worker.py:3548]: I'm a bit confused about the body of this handler. I don't think we should mix up the "CancelTask" event with a execute response that says "Didn't Execute. Task already cancelled"

File: distributed/worker.py [hunk 10]
```diff
@@ -3468,75 +3480,126 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
...
+        # key *must* be still in tasks. Releasing it directly is forbidden
+        # without going through cancelled
+        ts = self.tasks.get(ev.key)  # type: ignore
+        assert ts, self.story(ev.key)
+
+        ts.done = True
+        ts.startstops.append({"action": "compute", "start": ev.start, "stop": ev.stop})
+        ts.nbytes = ev.nbytes
+        ts.type = ev.type
+        return {ts: ("memory", ev.value)}, []
```
Comment (fjetter) [inline distributed/worker.py:3571]: We are loosing stimulus_ids here aren't we?

**Review by fjetter — COMMENTED** [review]
I think there is some confusion abut the CancelComputeEvent but otherwise this looks already good

File: distributed/worker.py [hunk 11]
```diff
@@ -2595,12 +2594,35 @@ def transitions(self, recommendations: Recs, *, stimulus_id: str) -> None:
         else:
             self._handle_instructions(instructions)
 
+    def handle_stimulus(self, stim: StateMachineEvent) -> None:
+        with log_errors():
+            # self.stimulus_history.append(stim)  # TODO
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/worker.py:2606]: Preferably no - it looks self-contained enough for a successive PR

File: distributed/worker.py [hunk 12]
```diff
@@ -3468,75 +3480,126 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
...
-            result["key"] = ts.key
-            value = result.pop("result", None)
-            ts.startstops.append(
-                {"action": "compute", "start": result["start"], "stop": result["stop"]}
-            )
-            self.threads[ts.key] = result["thread"]
-            recommendations: Recs = {}
+                self.active_keys.discard(key)
+
+            self.threads[key] = result["thread"]
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/worker.py:3492]: I understand that neither the Worker ABC nor the WorkerStateMachine will ever have a concept of threads. That's why I left it here.

File: distributed/worker.py (hunk 6 above)
Comment (crusaderky) [reply distributed/worker.py:3496]: self.digests is populated with a wealth of information that should remain alien to the state machine, so I think it should remain in Worker?

- latency
- transfer-bandwidth
- get-data-send-duration
- disk-load-duration
- profile-duration

File: distributed/worker.py (hunk 10 above)
Comment (crusaderky) [reply distributed/worker.py:3576]: We aren't; see handle_stimulus:
```python
            recs, instructions = self.handle_event(stim)
            self.transitions(recs, stimulus_id=stim.stimulus_id)
```

File: distributed/worker.py (hunk 8 above)
Comment (crusaderky) [reply distributed/worker.py:3560]: I had conflated the implementation of handle_cancel_compute coming from the scheduler and the code that dealt with a task being already cancelled by the time execute starts. They are now separate.

File: distributed/worker.py [hunk 13]
```diff
@@ -3468,75 +3480,126 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
...
+    @handle_event.register
+    def _(self, ev: CancelComputeEvent):  # -> tuple[Recs, Instructions]:
+        ts = self.tasks.get(ev.key)
+        if not ts or ts.state not in READY | {"waiting"}:
+            return {}, []
+
+        self.log.append((ev.key, "cancel-compute", ev.stimulus_id, time()))
+        # All possible dependents of ts should not be in state Processing on
+        # scheduler side and therefore should not be assigned to a worker, yet.
+        assert not ts.dependents
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/worker.py:3567]: I had conflated the implementation of handle_cancel_compute coming from the scheduler and the code that dealt with a task being already cancelled by the time execute starts. They are now separate.

[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/worker.py:3567]: It should now be functionally identical to main

File: distributed/worker.py (hunk 7 above)
Comment (crusaderky) [reply distributed/worker.py]: > Is this a functools limitation or why do we need py3.9? 

``@functools.singledispatchmethod`` exec()'s the delayed annotation - which however require Python 3.9+.
I worked around it better now.

> Do we even need individual annotations if they are all the same?

mypy is not smart enough to reserve a special treatment for singledispatch functions, so yes

File: distributed/worker.py (hunk 5 above)
Comment (crusaderky) [reply distributed/worker.py]: I split them now

**Comment by crusaderky** [comment]
@fjetter I did the work on _ensure_computing (see https://github.com/dask/distributed/pull/6003/commits/9d28bdeca23ab47776fffbcbb0a093e9b77801a2) but there are still some pretty conceptual failures caused by resource counting, so I'm moderately inclined to merge this PR as it is. I reverted it for now.
All tests should be green.

### In scope for this PR
- Remove all ``self.loop.add_callback(self.execute, ...)``
### Out of scope for the PR but in scope for the issue (to be implemented immediately after this PR)
- Refactor ``Worker.ensure_computing() -> None`` into ``Worker._ensure_computing() -> RecsInstr``
- Implement ``Worker.stimulus_history``
- Track tasks spawned by ``Worker._handle_instructions`` and cancel them in ``Worker.close``
### Out of scope for the issue
- Send Event Python objects from the Scheduler
- Other events (replicate etc.)

File: distributed/worker.py (hunk 6 above)
Comment (mrocklin) [reply distributed/worker.py:3501]: We can also nuke the digests if they're slowing us down.  They're not commonly used today.

**Review by fjetter — APPROVED** [review]
LGTM
